feat: Create the Streamy App

gvgesanto2 opened the pull request:
## Description

**Task:** https://github.com/microverseinc/curriculum-javascript/blob/main/group-capstone/js_capstone.md

This PR makes the following changes:
-  Creates the Streamy App using the TvMaze and Involvement APIs

Plus, it adopts the following methodologies and conventions:
- For the CSS/Sass:
  - [BEM](https://en.bem.info/methodology/) methodology for naming the CSS classes
  - A modified version of the [Scalable and Modular Architecture for CSS](http://smacss.com/) (SMACSS) methodology for organizing the CSS code into categorized folders/files (base, components, layout, and utilities)
- For the JavaScript:
  - [MVC architecture](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller) to separate the application business logic from the presentation logic (DOM manipulation tasks)
  
**Note: We changed the name of the ``comments`` popup to ``details`` popup to be more explicit about the purpose of this popup window (in our opinion, we believe that, as this popup shows the details information about the selected TV show, it makes more sense to call it ``details popup`` then ``comments popup``)**

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [X] Our code follows the style guidelines of this project (no linters errors)
- [X] We have checked my code and corrected any misspellings
- [X] We have commented our code
- [X] We have deleted any unused code as comments in our codebase

## Screenshots

- **Homepage in the Desktop version:**
![streamy-home-desktop](https://user-images.githubusercontent.com/64566209/175757030-090435ab-e698-42a5-866e-c9b397b4c941.png)
- **Homepage in the Mobile version:**
![streamy-home-mobile](https://user-images.githubusercontent.com/64566209/175757117-271a9bf5-384e-4535-805d-15ad7bda75a5.png)
- **Details popup in the Desktop version:**
![streamy-details-desktop](https://user-images.githubusercontent.com/64566209/175757130-58a7694b-06a2-4dbc-8fb9-657e1ed15680.png)
![streamy-details-desktop-2](https://user-images.githubusercontent.com/64566209/175757133-ed3cf8e2-325b-4988-8099-7c5c86af45e8.png)
![streamy-details-desktop-3](https://user-images.githubusercontent.com/64566209/175757137-aab13970-d19e-4418-bf4d-ed4af5534efe.png)
- **Details popup in the Mobile version:**
![streamy-details-mobile](https://user-images.githubusercontent.com/64566209/175757148-0e7c9dcd-958e-40b2-b265-344f85f7ca01.png)

gvgesanto2 committed:
feat: Set up Github actions, webpack and the base project structure

gvgesanto2 committed:
fix: Fix linters errors

gvgesanto2 committed:
Merge pull request #22 from gvgesanto2/setup

feat: Set up Github actions, Webpack and the base project structure

David-Verde committed:
add variables

David-Verde committed:
header

David-Verde committed:
cards

David-Verde committed:
cards

gvgesanto2 committed:
feat: Create the details popup basic structure and generate it dynamically

gvgesanto2 committed:
feat: Create the tv show description section dynamically

gvgesanto2 committed:
feat: Create the cast section dynamically

David-Verde committed:
cards2

gvgesanto2 committed:
fix: Fix linters errors

David-Verde committed:
API

gvgesanto2 committed:
Merge pull request #23 from gvgesanto2/details-popup

feat: Create the Details Popup, displaying details info about the selected TV Show

David-Verde committed:
mergin

David-Verde committed:
cards3

gvgesanto2 committed:
feat: Create the comments section dynamically

David-Verde committed:
test and like

David-Verde committed:
linters

gvgesanto2 committed:
feat: Finish touches in the comments and tv show cast sections

David-Verde committed:
Merge pull request #25 from gvgesanto2/homepage

Homepage

gvgesanto2 committed:
Solve merging conflicts

gvgesanto2 committed:
Merge pull request #26 from gvgesanto2/details-popup

feat: Create the Details Popup's Comments Section

David-Verde committed:
counter test fix

David-Verde committed:
Merge pull request #27 from gvgesanto2/test-counter

counter test fix

gvgesanto2 committed:
feat: Add the open details popup feature and add tests for the genCommentsCountText method

gvgesanto2 committed:
fix: Fix linters errors

gvgesanto2 committed:
Solve merging conflicts

gvgesanto2 committed:
docs: Update README file with the project description

gvgesanto2 committed:
Merge pull request #28 from gvgesanto2/test-comments-counter

Add tests for the comments counter

brytebee commented on README.md:
```diff
@@ -1 +1,94 @@
+  - MVC pattern
+  - BEM
+
+## Documentation Video
+
+Click on this link to see the documentation video: [Doc Video Link](https://drive.google.com/file/d/1vBo6ZyEETMqc5tm-bPsQLbM22L4Ev2Ev/view?usp=sharing)
+
```
- [ ] **Optional:** Awesome video presentation with the cool video editing 👏, however, it is longer than the specified length. In future presentations, endeavour to keep the specified length.

brytebee commented on src/scripts/showMovies.js:
```diff
@@ -0,0 +1,90 @@
+      );
+
+      const cast = await tvMazeApiService.getCastByShowId(data[i].id);
+      const comments = await involvementApiService.getCommentsById(data[i].id);
+      const detailsPopupController = new DetailsPopupController(
```
- [ ] **Required:** This call throws an error when a movie card without comments is opened. This is because you have not checked for cases when the movie has no comment. Kindly check for these cases, so that the fetch returns no error when the movie has no comments.

![image](https://user-images.githubusercontent.com/27709832/175760389-db5ba82c-950d-48f6-b3a8-c75fee96069f.png)

gvgesanto2 replied:
## Hi @brytebee,

Thank you for your feedback! However, **how can I check to see if a specific tv show has comments or not?** Maybe I'm missing something, but I couldn't find a way of doing that in the [Involvement API's documentation](https://www.notion.so/Involvement-API-869e60b5ad104603aa6db59e08150270). The only routes available in this API to deal with comments are these ones (at least for what I've read in the [Involvement API docs](https://www.notion.so/Involvement-API-869e60b5ad104603aa6db59e08150270)) :
- **GET** - /apps/**APP_ID**/comments?item_id=**SHOW_ID**: to get the comments available for the show indicated by the query param ``item_id`` equals to ``SHOW_ID``
- **POST** - /apps/**APP_ID**/comments?item_id=**SHOW_ID**: to create a new comment for the show indicated by the query param ``item_id`` equals to ``SHOW_ID``

**Therefore, when you open the popup for the first time, you have NO WAY to know if there are comments for the selected show.** So, you need to try to fetch this data from the API using the route ``GET - /apps/APP_ID/comments?item_id=SHOW_ID``. 
The problem is if there are no comments for the ``"SHOW_ID" show``, this request will throw an error, returning to the client an HTTP response with the ``status code 400``.

However, **this does NOT affect my application in any aspect (everything works fine)**, because if you check the ``InvolvementApiService.getCommentsById()`` method definition you will see that I'm actually handling this error inside a ``try-catch`` block:
![image](https://user-images.githubusercontent.com/64566209/175776388-304d944b-cbc7-4c18-9623-1d1785eccd05.png)

If the attempt to fetch the comments data for a specific show fails, I return an empty array, displaying the status message ``No Comments`` to the user.

If the problem is the fact this message is being shown in the Browser's console, I need to say that this is a default browser feature that is impossible to disable through JavaScript code (at least from what I have knowledge). You can check more about it in this StackOverflow post: https://stackoverflow.com/questions/51978983/how-to-remove-the-console-errors-in-axios
So, if there's no way to disable this feature in the user's browser through JavaScript, a palliative solution I could adopt is to clear the console every time I got this error from the API, like this:

![image](https://user-images.githubusercontent.com/64566209/175776981-0bba66b5-1ea1-4fc7-8232-7d46836b0fcc.png)

This will clear the browser's console after the error message mentioned is logged by the browser, preventing it to be displayed.

**Note: if there's a better way to approach this problem or if there's a route in the Involvement API to check if a specific show has any comments without returning an HTTP error response, I'd like to know.**

brytebee commented on src/scripts/views/view-components/details-popup/DetailsPopupView.js:
```diff
@@ -0,0 +1,117 @@
+    // Append the 'popupContent' children
+    const tvShowDescriptionSectionView = new TvShowDescriptionSectionView(this.tvShowData);
+    tvShowDescriptionSectionView.appendToParent(popupContent);
+    const tvShowCastSectionView = new TvShowCastSectionView(this.tvShowData, this.eventHandlersObj);
+    tvShowCastSectionView.appendToParent(popupContent);
+    this.commentsSectionView = new CommentsSectionView(id, comments, this.eventHandlersObj);
+    this.commentsSectionView.appendToParent(popupContent);
```
- [ ] **Required:** Kindly centre the elements in the comments section as in the wireframe.

Your Design | Template
:---: | :---:
![image](https://user-images.githubusercontent.com/27709832/175760535-c16bea75-1e7f-4352-b898-6a80e55c3e4c.png) | ![image](https://user-images.githubusercontent.com/27709832/175760572-6bb24754-416f-4aa9-b9e9-3f3235fd0913.png)

brytebee reviewed (changes requested):
Hi, @gvgesanto2 & @David-Verde👋, 

### Highlights 👏 
- The app fetches data from the API and renders the same to the page.
- The Involvement API has been correctly implemented.
- You have made some cool features, for example adding the cast and linking them to a more details page.
- All tests passes and green.
Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!

## Suggested changes
_Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._

Please, remember to tag me in your question so I can receive the notification. 
Please, remember to tag me @brytebee in your question so I can receive the notification.
### _**Use @brytebee on slack for faster responses.**_

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

gvgesanto2 committed:
feat: Center the elements in the details popup comments section

gvgesanto2 committed:
Merge pull request #30 from gvgesanto2/test-comments-counter

feat: Center the elements in the details popup comments section

gvgesanto2 commented:
## Description

**Task:** https://github.com/microverseinc/curriculum-javascript/blob/main/todo-list/m3_add_remove.md

This PR addresses the following changes requested by [the last reviewer](https://github.com/microverseinc/curriculum-javascript/blob/main/group-capstone/js_capstone.md):

> * [ ]  **Required:** This call throws an error when a movie card without comments is opened. This is because you have not checked for cases when the movie has no comment. Kindly check for these cases, so that the fetch returns no error when the movie has no comments.
> 
> ![image](https://user-images.githubusercontent.com/27709832/175760389-db5ba82c-950d-48f6-b3a8-c75fee96069f.png)

**Solution:** the solution adopted is the one I described in [this comment](https://github.com/gvgesanto2/microverse-js-capstone/pull/29#discussion_r906689639).

> * [ ]  **Required:** Kindly centre the elements in the comments section as in the wireframe.

**Solution:** 
![streamy-new-comments](https://user-images.githubusercontent.com/64566209/175778900-aa9d45f4-6276-4afd-aab7-847fe9062da8.png)

## Checklist:

- [X] My code follows the style guidelines of this project (no linters errors)
- [X] I have checked my code and corrected any misspellings
- [X] I have commented my code
- [X] I have deleted any unused code as comments in my codebase

youmari reviewed (approved):
Hi @gvgesanto2 @brytebee , 

In my opinion, your project is ready for the presentation!  There is nothing else to say other than... it's time to merge it :shipit: 

- General Requirements 👌
- Video Presentation 🙌
- Design is outstanding 🚀

Good luck! 🎉 
![gif](https://media.giphy.com/media/3oz9ZE2Oo9zRC/giphy.gif)

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
